pkg/controller/acmechallenges/finalizer.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package acmechallenges
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	cmacme "github.com/cert-manager/cert-manager/pkg/apis/acme/v1"
+)
+
+type challengeUpdater interface {
+	updateOrApply(context.Context, *cmacme.Challenge) (*cmacme.Challenge, error)
+}
+
+// setFinalizerStep is responsible for adding a cleanup finalizer to a challenge.
+// This ensures that the challenge is not garbage collected before cert-manager
+// has a chance to clean up resources created for the challenge.
+// When the challenge is marked for deletion, another step cleans up any
+// deployed ("presented") resources and if successful, removes this finalizer
+// allowing the garbage collector to remove the challenge.
+type setFinalizerStep struct {
+	challengeUpdater
+	ch *cmacme.Challenge
+}
+
+// Required returns true if the finalizer is not found on the challenge.
+func (o *setFinalizerStep) Required() bool {
+	return !sets.NewString(o.ch.Finalizers...).Has(cmacme.ACMEFinalizer)
+}
+
+// Run adds the finalizer to the challenge and saves the change to the API
+// server.
+func (o *setFinalizerStep) Run(ctx context.Context) (err error) {
+	ch := o.ch.DeepCopy()
+	ch.Finalizers = append(ch.Finalizers, cmacme.ACMEFinalizer)
+	ch, err = o.updateOrApply(ctx, ch)
+	if err != nil {
+		return err
+	}
+	*o.ch = *ch
```

Comment:
Why do we update the `ch` field on this struct? Looking at the only call site for this, we don't do anything with it

Reply:
Because this step does an Update which increments the metadata.generation and then there's a deferred UpdateStatus which will fail with a "object has been modified" error unless it submits the challenge with the updated generation and finalizers.